riscv/src/result.rs

```diff
@@ -0,0 +1,44 @@
+use core::fmt;
+
+/// Convenience alias for the [Result](core::result::Result) type for the library.
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Represents error variants for the library.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Error {
+    /// Attempted out-of-bounds access.
+    OutOfBounds {
+        index: usize,
+        min: usize,
+        max: usize,
+    },
+    /// Unimplemented function or type.
+    Unimplemented,
+    /// Invalid field value.
+    InvalidValue {
+        field: &'static str,
+        value: usize,
+        bitmask: usize,
```

Comment:
Maybe we can use something like:

```rust
pub enum InvalidValue {
    OutOfBounds {
        bitmask: usize,
    },
    ... // other causes
}
```

Or maybe this is overcomplicating stuff...

Reply:
> Or maybe this is overcomplicating stuff...

I think so, the print out of the error tells the reader that an invalid field value exists, what the value is, and what the valid bitmask is.

Easy enough for the person debugging the error to convert the values to binary, and see what bits don't match up.

Also what would semantically differentiate an "OutOfBounds" from an "Unknown" (or other variant). They would all technically fall outside the acceptable bitmask, no?

Or do you mean "OutOfBounds" as "value with a bit set below minimum, and/or value with a bit set above maximum" and "Unknown" as a "value with invalid bits set within the bitmask, e.g. bitmask = 0b101, value = 0b111"?

Either way, I think a user can get basically the same information from what exists now. We may want to format the output as binary instead of the current hex, though.